rasa/graph_components/providers/domain_provider.py

```diff
@@ -0,0 +1,44 @@
+from __future__ import annotations
+from typing import Dict, Text, Any
+
+from rasa.engine.graph import GraphComponent, ExecutionContext
+from rasa.engine.storage.resource import Resource
+from rasa.engine.storage.storage import ModelStorage
+from rasa.shared.core.domain import Domain
+from rasa.shared.importers.importer import TrainingDataImporter
+
+
+class DomainProvider(GraphComponent):
```

Comment:
Reading the documentation I saw that the DomainProvider is marked as a persisting component. However, I also don't fully understand what should be persisted here and why ~ @wochinge is that persisting still relevant or did we remove that anyway by making the TrainingDataImporter fingerprint random?

Reply:
@twerkmeister Very good point! This component needs to be able to persist as this component also provides the domain during inference (our policies need the domain during inference). So the code should roughtly look like this:

```python

class DomainProvider:

   def __init__(domain: Optional):
    ...
    
   def provide_train(self, importer):
      domain = importer.get_domain()
      self.persist(self._resource, self._model_storage, domain)
      return domain
      
   def load(...):
     with model_storage(...):
        domain = read_domain_from_file)
     return cls(domain)
     
   def provide_inference(self):
      return self.domain
```

Reply:
@wochinge which approach should I take for `persist` method 🤔 :
```
with self._model_storage.write(...) as directory:
        self.domain.persist(directory)
```
OR

```
domain_yaml = self.domain.as_yaml()
self._resource.to_cache(Path(domain_yaml), self._model_storage)
```

Reply:
the first one please 🙌🏻 

The second one should ideally never be called manually. That's done by the the `LocalTrainingCache` in the `TrainingHook`